src/coprocessor/cache.rs

```diff
@@ -0,0 +1,487 @@
+            let region_id: u64 = entry.region_id;
+            if let Some(node) = regions.get_mut(&region_id) {
+                // Delete from region cache entry list
+                // We should not delete self.regions item.
+                // In some coner case delete self.regions item may cause
+                // cache no need data and then cause an error result.
+                node.cached_items.remove(k);
+            }
+            self.size -= entry.size;
+        }
+    }
+
+    fn evict_region_with_exist_region(&mut self, region_id: u64) {
+        self.regions
+            .get_mut(&region_id)
+            .map_or(None, |region| {
+                region.version += 1;
+                Some(region.cached_items.clone())
+            })
+            .map_or(None, |cached_items| {
+                for (key, _) in (&cached_items).iter() {
+                    self.remove(key);
+                }
+                Some(())
+            });
+    }
+
+    fn evict_region_with_new_region(&mut self, region_id: u64) {
+        // If there's no region version data in cache, get_version will
+        // return 0 and after that some thread call evict_region, we should set
+        // version to 1 to prevent cache unused data.
+        let entry = RegionDistSQLCacheEntry {
+            version: 1,
+            ..Default::default()
+        };
+        self.regions.insert(region_id, entry);
+    }
+
+    pub fn evict_region(&mut self, region_id: u64) {
+        if self.regions.contains_key(&region_id) {
+            self.evict_region_with_exist_region(region_id);
+        } else {
+            self.evict_region_with_new_region(region_id);
+        }
+
+        CORP_DISTSQL_CACHE_SIZE_GAUGE_VEC
+            .with_label_values(&["size"])
+            .set(self.size() as f64);
+
+        CORP_DISTSQL_CACHE_SIZE_GAUGE_VEC
+            .with_label_values(&["count"])
+            .set(self.len() as f64);
+    }
+
+    pub fn evict_region_and_enable(&mut self, region_id: u64) {
+        self.evict_region(region_id);
+        self.enable_region_cache(region_id);
+    }
+
+    pub fn capacity(&self) -> usize {
+        self.max_size
+    }
+
+    pub fn len(&self) -> usize {
+        self.map.len()
+    }
+
+    pub fn size(&self) -> usize {
+        self.size
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.map.len() == 0
+    }
+
+    pub fn update_capacity(&mut self, capacity: usize) {
+        self.max_size = capacity;
+
+        // Remove entry untile cache size is less or equals than capacity
+        while self.size() > self.capacity() {
+            self.remove_lru();
+        }
+    }
+
+    pub fn disable_region_cache(&mut self, region_id: u64) {
+        let mut entry = self.regions
+            .entry(region_id)
+            .or_insert_with(Default::default);
+        entry.enable = false;
+    }
+
+    pub fn enable_region_cache(&mut self, region_id: u64) {
+        self.regions.get_mut(&region_id).and_then(|e| {
```

Comment:
Could we use `map` instead of `and_then`?

Reply:
As I think, and_then is more meaningful than map. This code can be translate as:

Reply:
I think `map` and `and_then` have the same meaning in this situation--apply a function when the option is something, meanwhile, you can write less code  by using map:
```
self.regions.get_mut(&region_id).map(|e|e.enable=true);
```